src/readers/src/virtex_data_provider.cpp

```diff
@@ -7,27 +7,38 @@ Box3D liftBox(const Box3D box, float offset) {
 	coordinates.back().second += offset;
 	return {coordinates};
 }
-
 }  // namespace
 
 namespace Readers {
-VirtexDataProvider::VirtexDataProvider(const TRM& geom,
-                                       const T2D& heat) {
+VirtexDataProvider::VirtexDataProvider(const TRM& geom, const T2D& heat) {
 	load_geometry(geom);
 	load_heatmap(heat);
 	load_metadata();
 }
+
 void VirtexDataProvider::load_geometry(const TRM& data) {
-	float offset = 0;
+	float offset          = 0;
+	size_t box_counter_   = 0;
+	size_t layer_counter_ = 0;
 	geometry_.Clear();
-	for (const auto& layer : data.layers_) {
-		auto geom = layer->geometry();
-		for (const auto& shape : geom.get_all_shapes()) {
-			shape->setBox(liftBox(shape->bbox(), offset));
-			offset += layer->thickness();
-			geometry_.AddShape(shape);
+	for (const auto& [position, layers] : data.layers_groups_) {
+		for (const auto& layer : layers) {
+			const auto shapes = layer->geometry().get_all_shapes();
+			for (const auto& shape : shapes) {
```

Comment:
Why can't we use `for (const auto& shape : geom.get_all_shapes())`?

Reply:
For future PR